File: raglift/citations.py
from langchain_core.documents import Document

from raglift.schemas import Source


def build_sources(documents: list[Document], preview_chars: int = 180) -> list[Source]:
    sources: list[Source] = []
    seen: set[str] = set()
    for doc in documents:
        chunk_id = str(doc.metadata.get("chunk_id", ""))
        path = str(doc.metadata.get("path", doc.metadata.get("source", "")))
        key = f"{path}:{chunk_id}"
        if key in seen:
            continue
        seen.add(key)
        sources.append(
            Source(
                path=path,
                chunk_id=chunk_id,
                content_preview=doc.page_content[:preview_chars].replace("\n", " "),
            )
        )
    return sources

File: src/citations.ts
import type { Source } from "./types.js";

type CitationDocument = {
  pageContent: string;
  metadata: Record<string, unknown>;
};

export function buildSources(documents: CitationDocument[], previewChars = 180): Source[] {
  const sources: Source[] = [];
  const seen = new Set<string>();
  for (const doc of documents) {
    const chunkId = String(doc.metadata.chunk_id ?? doc.metadata.chunkId ?? "");
    const path = String(doc.metadata.path ?? doc.metadata.source ?? "");
    const key = `${path}:${chunkId}`;
    if (seen.has(key)) continue;
    seen.add(key);
    sources.push({
      path,
      chunkId,
      contentPreview: doc.pageContent.slice(0, previewChars).replace(/\n/g, " "),
    });
  }
  return sources;
}
